From b9ee13361a920328eabb7639a768e2b13eec405d Mon Sep 17 00:00:00 2001
From: Timothy Arceri <tarceri@itsqueeze.com>
Date: Fri, 22 Jun 2018 18:47:51 +1000
Subject: [PATCH 03/12] mesa: enable ARB_gpu_shader_fp64 in compat profile

---
 src/mesa/main/extensions_table.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/mesa/main/extensions_table.h b/src/mesa/main/extensions_table.h
index 7af48a4ad9..7f472ee8c0 100644
--- a/src/mesa/main/extensions_table.h
+++ b/src/mesa/main/extensions_table.h
@@ -76,7 +76,7 @@ EXT(ARB_get_program_binary                  , dummy_true
 EXT(ARB_get_texture_sub_image               , dummy_true                             , GLL, GLC,  x ,  x , 2014)
 EXT(ARB_gl_spirv                            , ARB_gl_spirv                           ,  x,  GLC,  x ,  x , 2016)
 EXT(ARB_gpu_shader5                         , ARB_gpu_shader5                        , GLL, GLC,  x ,  x , 2010)
-EXT(ARB_gpu_shader_fp64                     , ARB_gpu_shader_fp64                    ,  x , GLC,  x ,  x , 2010)
+EXT(ARB_gpu_shader_fp64                     , ARB_gpu_shader_fp64                    , GLL, GLC,  x ,  x , 2010)
 EXT(ARB_gpu_shader_int64                    , ARB_gpu_shader_int64                   ,  x , GLC,  x ,  x , 2015)
 EXT(ARB_half_float_pixel                    , dummy_true                             , GLL, GLC,  x ,  x , 2003)
 EXT(ARB_half_float_vertex                   , ARB_half_float_vertex                  , GLL, GLC,  x ,  x , 2008)
-- 
2.18.0

