diff --git a/src/vulkan/overlay-layer/VkLayer_MESA_overlay.json.in b/src/vulkan/overlay-layer/VkLayer_MESA_overlay.json.in
index 5a9a88a98416d262b2e3d125b24d22b4b6e65066..6877c3a3ecdf4f5bf65a3e2609bafc087164f685 100644
--- a/src/vulkan/overlay-layer/VkLayer_MESA_overlay.json.in
+++ b/src/vulkan/overlay-layer/VkLayer_MESA_overlay.json.in
@@ -3,7 +3,7 @@
     "layer" : {
         "name": "VK_LAYER_MESA_overlay",
         "type": "GLOBAL",
-        "library_path": "@install_dir@/libVkLayer_MESA_overlay.so",
+        "library_path": "libVkLayer_MESA_overlay.so",
         "api_version": "1.1.73",
         "implementation_version": "1",
         "description": "Mesa Overlay layer"
diff --git a/src/vulkan/overlay-layer/meson.build b/src/vulkan/overlay-layer/meson.build
index 9d1bc97c13153de54d075b8af33f981e2b9815a3..a8f0caf9db99cdaac77f6f81e1bbe621e80597a4 100644
--- a/src/vulkan/overlay-layer/meson.build
+++ b/src/vulkan/overlay-layer/meson.build
@@ -18,8 +18,6 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
 
-vklayer_install_dir = join_paths(get_option('prefix'), 'share', 'vulkan', 'explicit_layer.d')
-
 glslang = find_program('glslangValidator')
 
 overlay_shaders = [
@@ -47,17 +45,15 @@ vklayer_mesa_overlay = shared_library(
   include_directories : [inc_common, inc_vulkan_util],
   link_args : cc.get_supported_link_arguments(['-Wl,-Bsymbolic-functions', '-Wl,-z,relro']),
   link_with : [libmesa_util, libvulkan_util],
-  install_dir : vklayer_install_dir,
   install : true
 )
 
-sdata = configuration_data()
-sdata.set('install_dir', vklayer_install_dir)
+vklayer_install_dir = join_paths(get_option('prefix'), 'share', 'vulkan', 'explicit_layer.d')
 
 configure_file(
   input : 'VkLayer_MESA_overlay.json.in',
   output : '@BASENAME@',
-  configuration : sdata,
+  configuration : configuration_data(),
   install_dir : vklayer_install_dir,
   install : true
 )
