diff --git a/src/loader/pci_id_driver_map.h b/src/loader/pci_id_driver_map.h
index f5fc349c1cc..581d53295a7 100644
--- a/src/loader/pci_id_driver_map.h
+++ b/src/loader/pci_id_driver_map.h
@@ -66,7 +66,7 @@ static const struct {
    bool (*predicate)(int fd);
 } driver_map[] = {
    { 0x8086, "i915", i915_chip_ids, ARRAY_SIZE(i915_chip_ids) },
-   { 0x8086, "i965", i965_chip_ids, ARRAY_SIZE(i965_chip_ids) },
+   { 0x8086, "iris", i965_chip_ids, ARRAY_SIZE(i965_chip_ids) },
    { 0x8086, "iris", NULL, -1 },
    { 0x1002, "radeon", r100_chip_ids, ARRAY_SIZE(r100_chip_ids) },
    { 0x1002, "r200", r200_chip_ids, ARRAY_SIZE(r200_chip_ids) },
