From 691f1823eccf0e8ff1f676a3f0a3a4a7b12320eb Mon Sep 17 00:00:00 2001
From: Mike Lothian <mike@fireburn.co.uk>
Date: Tue, 13 Feb 2024 20:29:56 +0000
Subject: [PATCH] Revert "meson: enforce build of intel-clc with anv/iris"

This reverts commit a512c2a8b572c5da360873320dbbd343c6223cd6.

diff --git a/meson.build b/meson.build
index faadf25a449..ade6546b0af 100644
--- a/meson.build
+++ b/meson.build
@@ -295,7 +295,8 @@ with_any_broadcom = [
   with_broadcom_vk,
 ].contains(true)
 
-if get_option('intel-clc') != 'system'
+if ['x86_64'].contains(host_machine.cpu_family()) and \
+   get_option('intel-clc') != 'system'
   # Require intel-clc with Anv & Iris (for internal shaders)
   with_intel_clc = get_option('intel-clc') == 'enabled' or \
                    with_intel_vk or with_gallium_iris
@@ -303,7 +304,9 @@ else
   with_intel_clc = false
 endif
 
-with_intel_vk_rt = with_intel_vk and get_option('intel-rt') != 'disabled'
+with_intel_vk_rt = with_intel_vk and \
+                   get_option('intel-clc') != 'disabled' and \
+                   get_option('intel-rt') != 'disabled'
 
 with_any_intel = [
   with_gallium_crocus,
diff --git a/meson_options.txt b/meson_options.txt
index ae419fce5dc..5311348719b 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -635,11 +635,11 @@ option(
   'intel-clc',
   type : 'combo',
   deprecated: {'true': 'enabled', 'false': 'disabled'},
-  value : 'auto',
+  value : 'disabled',
   choices : [
-    'enabled', 'system', 'auto'
+    'enabled', 'disabled', 'system',
   ],
-  description : 'Build the intel-clc compiler or use a system version.'
+  description : 'Build the intel-clc compiler.'
 )
 
 option(
-- 
2.43.1

