From 313178acd1f847765587635be76ad20f5b6d67c3 Mon Sep 17 00:00:00 2001
From: Mike Lothian <mike@fireburn.co.uk>
Date: Sat, 10 Feb 2024 16:03:38 +0000
Subject: [PATCH] Fix for warned

---
 src/mesa/main/glformats.c | 11 -----------
 1 file changed, 11 deletions(-)

diff --git a/src/mesa/main/glformats.c b/src/mesa/main/glformats.c
index 28d7b998433..a832aa42cb4 100644
--- a/src/mesa/main/glformats.c
+++ b/src/mesa/main/glformats.c
@@ -2902,17 +2902,6 @@ _mesa_gles_check_internalformat(struct gl_context *ctx,
       return GL_NO_ERROR;
 
    case GL_BGRA8_EXT:
-      /* This is technically speaking out-of-spec. But too many
-       * applications seems to depend on it, so let's allow it
-       * together with a small complaint */
-      static bool warned = false;
-      if (!warned) {
-         _mesa_warning(ctx,
-            "internalformat = GL_BGRA8_EXT invalid by spec, but too many "
-            "applications depend on it to error. Plese fix the software "
-            "that causes this problem.");
-         warned = true;
-      }
       return GL_NO_ERROR;
 
    default:
-- 
2.43.0

