diff --git a/src/gallium/frontends/rusticl/meson.build b/src/gallium/frontends/rusticl/meson.build
index 39b2089fcfc..0663a2b6c59 100644
--- a/src/gallium/frontends/rusticl/meson.build
+++ b/src/gallium/frontends/rusticl/meson.build
@@ -347,6 +347,7 @@ rusticl_c = static_library(
   ],
   gnu_symbol_visibility : 'hidden',
   include_directories : [
+    meson.build_root(),
     inc_gallium,
     inc_gallium_aux,
     inc_include,
