diff --git a/src/amd/llvm/ac_llvm_util.c b/src/amd/llvm/ac_llvm_util.c
index 575b6e45567..ebed019c96f 100644
--- a/src/amd/llvm/ac_llvm_util.c
+++ b/src/amd/llvm/ac_llvm_util.c
@@ -65,9 +65,6 @@ static void ac_init_llvm_target(void)
     */
    const char *argv[] = {
       "mesa",
-      "-simplifycfg-sink-common=false",
-      "-global-isel-abort=2",
-      "-amdgpu-atomic-optimizations=true",
 #if LLVM_VERSION_MAJOR == 11
       /* This fixes variable indexing on LLVM 11. It also breaks atomic.cmpswap on LLVM >= 12. */
       "-structurizecfg-skip-uniform-regions",
