From 98dfa2fa8158f90c8efafdce1232892f5fe53188 Mon Sep 17 00:00:00 2001
From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Thu, 6 Feb 2020 23:43:02 +0900
Subject: [PATCH] repart: quit earlier if no .conf file exists

Fixes #14804.
---
 src/partition/repart.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/src/partition/repart.c b/src/partition/repart.c
index 58fb9c7c5d3..71ac2bc0197 100644
--- a/src/partition/repart.c
+++ b/src/partition/repart.c
@@ -2938,10 +2938,6 @@ static int run(int argc, char *argv[]) {
         if (r < 0)
                 return r;
 
-        r = find_root(&node);
-        if (r < 0)
-                return r;
-
         context = context_new(arg_seed);
         if (!context)
                 return log_oom();
@@ -2950,6 +2946,13 @@ static int run(int argc, char *argv[]) {
         if (r < 0)
                 return r;
 
+        if (context->n_partitions <= 0)
+                return 0;
+
+        r = find_root(&node);
+        if (r < 0)
+                return r;
+
         r = context_load_partition_table(context, node);
         if (r == -EHWPOISON)
                 return 77; /* Special return value which means "Not GPT, so not doing anything". This isn't
