diff -ru '--exclude=*.o' ../minidlna-1.2.1-orig/image_utils.c ./image_utils.c
--- a/image_utils.c	2020-05-23 13:18:05.155000930 +0100
+++ b/image_utils.c	2020-05-23 13:37:22.343534082 +0100
@@ -190,7 +190,7 @@
 	src->pub.bytes_in_buffer = bufsize;
 }
 
-jmp_buf setjmp_buffer;
+static jmp_buf setjmp_buffer;
 /* Don't exit on error like libjpeg likes to do */
 static void
 libjpeg_error_handler(j_common_ptr cinfo)
--- a/metadata.c	2020-05-23 13:18:05.155000930 +0100
+++ b/metadata.c	2020-05-23 13:37:13.719641896 +0100
@@ -484,7 +484,7 @@
 }
 
 /* For libjpeg error handling */
-jmp_buf setjmp_buffer;
+static jmp_buf setjmp_buffer;
 static void
 libjpeg_error_handler(j_common_ptr cinfo)
 {
--- a/tivo_utils.c	2020-05-23 13:18:05.164000817 +0100
+++ b/tivo_utils.c	2020-05-23 13:29:56.778104417 +0100
@@ -27,6 +27,7 @@
 #include <sqlite3.h>
 #include "tivo_utils.h"
 
+struct sqlite3PrngType sqlite3Prng;
 /* This function based on byRequest */
 char *
 decodeString(char *string, int inplace)
--- a/tivo_utils.h	2020-05-23 13:18:05.164000817 +0100
+++ b/tivo_utils.h	2020-05-23 13:30:07.354972188 +0100
@@ -30,7 +30,9 @@
   unsigned char isInit;          /* True if initialized */
   unsigned char i, j;            /* State variables */
   unsigned char s[256];          /* State variables */
-} sqlite3Prng;
+};
+
+extern struct sqlite3PrngType sqlite3Prng;
 
 char *
 decodeString(char *string, int inplace);
