From 284e5a6e9f3b1aec2571f124c89f79a76cec7790 Mon Sep 17 00:00:00 2001
From: Thong Thai <thong.thai@amd.com>
Date: Wed, 16 Mar 2022 14:16:15 -0400
Subject: [PATCH] change blit for rgb

---
 src/gallium/frontends/va/postproc.c | 6 +-----
 1 file changed, 1 insertion(+), 5 deletions(-)

diff --git a/src/gallium/frontends/va/postproc.c b/src/gallium/frontends/va/postproc.c
index a9c76f4039f..dc8bae7da33 100644
--- a/src/gallium/frontends/va/postproc.c
+++ b/src/gallium/frontends/va/postproc.c
@@ -400,11 +400,7 @@ vlVaHandleVAProcPipelineParameterBufferType(vlVaDriver *drv, vlVaContext *contex
 
    if (context->target->buffer_format != PIPE_FORMAT_NV12 &&
        context->target->buffer_format != PIPE_FORMAT_P010 &&
-       context->target->buffer_format != PIPE_FORMAT_P016 &&
-       context->target->buffer_format != PIPE_FORMAT_B8G8R8X8_UNORM &&
-       context->target->buffer_format != PIPE_FORMAT_B8G8R8A8_UNORM &&
-       context->target->buffer_format != PIPE_FORMAT_R8G8B8X8_UNORM &&
-       context->target->buffer_format != PIPE_FORMAT_R8G8B8A8_UNORM)
+       context->target->buffer_format != PIPE_FORMAT_P016)
       return vlVaPostProcCompositor(drv, context, src_region, dst_region,
                                     src, context->target, deinterlace);
    else
-- 
2.35.1

