From 45930c059541a493c63208abb753653f797167f8 Mon Sep 17 00:00:00 2001
From: Dave Airlie <airlied@redhat.com>
Date: Mon, 14 Mar 2022 04:49:42 +1000
Subject: [PATCH] add some debug

---
 src/vulkan/wsi/wsi_common.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/vulkan/wsi/wsi_common.c b/src/vulkan/wsi/wsi_common.c
index 5c89e7ceb1f..f7dd531c3a5 100644
--- a/src/vulkan/wsi/wsi_common.c
+++ b/src/vulkan/wsi/wsi_common.c
@@ -364,6 +364,8 @@ wsi_configure_image(const struct wsi_swapchain *chain,
    memset(info, 0, sizeof(*info));
    uint32_t *queue_family_indices;
 
+   fprintf(stderr, "image sharing: %d, %d, %p\n", pCreateInfo->imageSharingMode,
+           pCreateInfo->queueFamilyIndexCount, pCreateInfo->pQueueFamilyIndices);
    if (pCreateInfo->imageSharingMode == VK_SHARING_MODE_CONCURRENT) {
       queue_family_indices =
          vk_alloc(&chain->alloc,
-- 
2.35.1

