diff --git a/tests/Makefile b/tests/Makefile
index dcb89ae..f1ea608 100644
--- a/tests/Makefile
+++ b/tests/Makefile
@@ -7,12 +7,12 @@ include ../Make.Rules
 
 DEPS=../libcap/libcap.a ../libcap/libpsx.a
 
-all: psx_test psx_test_wrap libcap_psx_test
+all: psx_test
 
 $(DEPS):
 	make -C ../libcap all
 
-test: run_psx_test run_libcap_psx_test
+test: run_psx_test
 
 sudotest: test
 
@@ -20,21 +20,11 @@ CFLAGS += -fPIC
 
 install: all
 
-run_psx_test: psx_test psx_test_wrap
+run_psx_test: psx_test
 	./psx_test
-	./psx_test_wrap
 
 psx_test: psx_test.c $(DEPS)
 	$(CC) $(CFLAGS) $(CPPFLAGS) -DNOWRAP $< -o $@ $(LIBPSXLIB)
 
-psx_test_wrap: psx_test.c $(DEPS)
-	$(CC) $(CFLAGS) $(CPPFLAGS) $< -o $@ $(LIBPSXLIB) -Wl,-wrap,pthread_create
-
-run_libcap_psx_test: libcap_psx_test
-	./libcap_psx_test
-
-libcap_psx_test: libcap_psx_test.c $(DEPS)
-	$(CC) $(CFLAGS) $(CPPFLAGS) $< -o $@ $(LIBCAPLIB) $(LIBPSXLIB) -Wl,-wrap,pthread_create --static
-
 clean:
-	rm -f psx_test psx_test_wrap libcap_psx_test
+	rm -f psx_test
