From 5cdc65b2991954726eee3309ecd67a6be38346ac Mon Sep 17 00:00:00 2001
From: Mike Lothian <mike@fireburn.co.uk>
Date: Thu, 31 Aug 2023 00:23:52 +0100
Subject: [PATCH] Fix versions

---
 CMakeLists.txt | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1e2260b09..713c98295 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -316,7 +316,7 @@ find_package(ZLIB 1.2 REQUIRED)
 find_package(zstd 1.5 REQUIRED)
 
 if (NOT YUZU_USE_EXTERNAL_VULKAN_HEADERS)
-    find_package(Vulkan 1.3.274 REQUIRED)
+    find_package(Vulkan 1.3.250 REQUIRED)
 endif()
 
 if (ENABLE_LIBUSB)
@@ -341,7 +341,7 @@ endif()
 
 if (ENABLE_WEB_SERVICE)
     find_package(cpp-jwt 1.4 CONFIG)
-    find_package(httplib 0.12 MODULE COMPONENTS OpenSSL)
+    find_package(httplib 0.14 MODULE COMPONENTS OpenSSL)
 endif()
 
 if (YUZU_TESTS)
-- 
2.42.0

