diff --git a/src/core/ksambashare.cpp b/src/core/ksambashare.cpp
index 7022c442e..d105ad17c 100644
--- a/src/core/ksambashare.cpp
+++ b/src/core/ksambashare.cpp
@@ -192,21 +192,8 @@ QByteArray KSambaSharePrivate::getNetUserShareInfo()
         QStringLiteral("info"),
     };
 
-    runProcess(QStringLiteral("net"), args, stdOut, stdErr);
 
-    if (!stdErr.isEmpty()) {
-        if (stdErr.contains("You do not have permission to create a usershare")) {
-            skipUserShare = true;
-        } else if (stdErr.contains("usershares are currently disabled")) {
-            skipUserShare = true;
-        } else {
-            // TODO: parse and process other error messages.
-            // create a parser for the error output and
-            // send error message somewhere
-            qCDebug(KIO_CORE) << "We got some errors while running 'net usershare info'";
-            qCDebug(KIO_CORE) << stdErr;
-        }
-    }
+    skipUserShare = true;
 
     return stdOut;
 }
