diff --git a/dlls/ntdll/unix/virtual.c b/dlls/ntdll/unix/virtual.c
index 4062921447..0a41787dea 100644
--- a/dlls/ntdll/unix/virtual.c
+++ b/dlls/ntdll/unix/virtual.c
@@ -575,9 +575,18 @@ static void free_ranges_remove_range( void *view_base, void *view_end, void *vie
         return;
     }
 
+    if (range->end == view_base && next->base > view_base && next->end >= view_end)
+    {
+        WARN( "range %p - %p is already partially mapped %p - %p (next %p - %p)\n", view_base, view_end, range->base, range->end, next->base, next->end );
+        view_base = next->base;
+        range = next;
+        next = range + 1;
+        assert( range->end > view_base || next != free_ranges_end );
+    }
+
     /* this should never happen */
     if (range->base > view_base || range->end < view_end)
-        ERR( "range %p - %p is already partially mapped\n", view_base, view_end );
+        ERR( "range %p - %p is already partially mapped %p - %p (next %p - %p)\n", view_base, view_end, range->base, range->end, next->base, next->end );
     assert( range->base <= view_base && range->end >= view_end );
 
     /* need to split the range in two */
