diff --git a/hw/xfree86/dri2/dri2.c b/hw/xfree86/dri2/dri2.c
index 6619e3aa7..8c9ba2518 100644
--- a/hw/xfree86/dri2/dri2.c
+++ b/hw/xfree86/dri2/dri2.c
@@ -1445,11 +1445,11 @@ dri2_probe_driver_name(ScreenPtr pScreen, DRI2InfoPtr info)
     char *driver = NULL;
     drmDevicePtr dev;
 
-    /* For non-PCI devices and drmGetDevice fail, just assume that
+    /* For non-PCI devices and drmGetDevice2 fail, just assume that
      * the 3D driver is named the same as the kernel driver. This is
      * currently true for vc4 and msm (freedreno).
      */
-    if (drmGetDevice(info->fd, &dev) || dev->bustype != DRM_BUS_PCI) {
+    if (drmGetDevice2(info->fd, 0, &dev) || dev->bustype != DRM_BUS_PCI) {
         drmVersionPtr version = drmGetVersion(info->fd);
 
         if (!version) {
