diff --git a/present/present_wnmd.c b/present/present_wnmd.c
index 72bda2f..80ffb01 100644
--- a/present/present_wnmd.c
+++ b/present/present_wnmd.c
@@ -518,8 +518,6 @@ present_wnmd_window_to_crtc_msc(WindowPtr window, RRCrtcPtr crtc, uint64_t windo
     present_window_priv_ptr window_priv = present_get_window_priv(window, TRUE);
 
     if (crtc != window_priv->crtc) {
-        uint64_t old_ust, old_msc;
-
         if (window_priv->crtc == PresentCrtcNeverSet) {
             window_priv->msc_offset = 0;
         } else {
@@ -527,10 +525,7 @@ present_wnmd_window_to_crtc_msc(WindowPtr window, RRCrtcPtr crtc, uint64_t windo
              * we'll just use whatever previous MSC we'd seen from this CRTC
              */
 
-            if (present_wnmd_get_ust_msc(window->drawable.pScreen, window, &old_ust, &old_msc) != Success)
-                old_msc = window_priv->msc;
-
-            window_priv->msc_offset += new_msc - old_msc;
+            window_priv->msc_offset += new_msc - window_priv->msc;
         }
         window_priv->crtc = crtc;
     }
